### DISCUSSION TITLE
Question... Are some dependencies missing from Umbraco 9.5?

### DISCUSSION THREAD

**Discussion by hfloyd** [umbraco/Umbraco-CMS#12383] [Features and ideas]
I set up a fresh 9.5 site using the [VS "Umbraco Project" template](https://our.umbraco.com/documentation/Fundamentals/Setup/Install/install-umbraco-with-templates), and came across an odd runtime issue when using IIS for the localhost (because I like to use real hostnames and not always run sites in debug-mode via VS).

@NikRimington summed up the issue most succinctly when I did a search on the Umbraco Discord about it, so I will quote him:

> Does anyone know why Umbraco v9 in IIS results in a "InvalidOperationException: Cannot find compilation library location for package 'System.Drawing.Common'" but running it via Kestral doesn't?

My experience was the same - if I ran it via IIS Express (debug inside VS) - no runtime error. If I used IIS, pointing to the UmbracoProject folder - Runtime error.

I didn't see any specific responses to Nik's query, and I went on my own wild-goose-chase to figure out if there was some magical setting which would make it "just work".... (After all, I was running an Umbraco 9 Cloud site locally without this issue...) Alas, too many hours later, I just ended up adding a reference via NuGet for "System.Drawing.Common" Version="5.0.3", which resolved that issue, but brought up a new, similar error about missing "System.Security.Cryptography.Pkcs", so I added that as well ("System.Security.Cryptography.Pkcs" Version="5.0.1"), and now the site loads via IIS. 

So, my question is : **Are these required references which are just missing from the Umbraco.Cms NuGet package?** If not, does anyone have an understanding of what is going on here? 

A few random (maybe related?) references:
[System.Drawing.Common.dll is not added to Net Core build output](https://github.com/dotnet/sdk/issues/11489#)

[PackageReference assembly is not included in the bin dir of a .NET standard project](https://github.com/dotnet/sdk/issues/747#)

[InvalidOperationException: Cannot find compilation library location for package 'System.Security.Cryptography.Pkcs'](https://stackoverflow.com/questions/70236519/invalidoperationexception-cannot-find-compilation-library-location-for-package)

[Razor templates: Cannot find compilation library location for package](https://stackoverflow.com/questions/51269559/razor-templates-cannot-find-compilation-library-location-for-package)

**Comment by nul800sebastiaan** [comment]
Which .NET version are you on? And @NikRimington ?

We should get better at promoting this button for these kinds of questions 😅 

![image](https://user-images.githubusercontent.com/304656/167571834-921d31f5-4eb5-4279-8ea6-9f68306f2640.png)

Which results in something like:

<html>
<body>


Category | Data
-- | --
Server OS | Microsoft Windows 10.0.14393
Server Framework | .NET 5.0.15
Default Language | en-US
Umbraco Version | 9.3.2-apreview001
Current Culture | en-US
Current UI Culture | en-US
Current Webserver | IIS
Browser | Edge-chromium 101.0.1210.39
Browser OS | Win32


</body>
</html>

We don't rely on `System.Drawing.Common` ourselves, though Imagesharp seems to need it but only for tests.

I can see I don't have the dll for a dotnet 5 Umbraco site (anecdotally, I only tried one) and it is copied into the bin for dotnet 6 Umbraco sites (only looked at 2 of them).

Do you have both the sdk and the hosting bundle installed on the machine?

Hosting bundle: https://dotnet.microsoft.com/en-us/download/dotnet/6.0/runtime
SDK: https://dotnet.microsoft.com/en-us/download

**Reply by NikRimington** [reply]
To be honest, I've abandoned IIS for the time being and I'm purely using IIS Express / Kestral when running the sites locally. So any comments I add are from memory, but I do remember installing the hosting bundle and the SDK. 

I recently did the Load Balancing Course and had a v9 site running in IIS with no problems and I didn't have to install either (so making the assumption that both are already installed).

For a time frame, my discord post is dated 2022-03-02 so it's a couple of months old.


In simple, I never solved the problem as I wasn't prepared to install "unnecessary" packages.  But I have found that sometimes I have to do a dotnet build before a dotnet run or dotnet publish for Nuget to restore things correctly.

**Reply by hfloyd** [reply]
My site info:
<html>
<body>


Category | Data
-- | --
Server OS | Microsoft Windows 10.0.19044
Server Framework | .NET 5.0.16
Default Language | en-US
Umbraco Version | 9.5.0
Current Culture | en-US
Current UI Culture | en-US
Current Webserver | IIS
Models Builder Mode | Nothing
Debug Mode | True
Database Provider | System.Data.SqlClient
Browser | Chrome 101.0.4951.54
Browser OS | Win32


</body>
</html>

(Wow - I didn't know that was there! Learning new things every day! Thanks, @nul800sebastiaan 😊)

**Comment by hfloyd** [comment]
RE: 

> In simple, I never solved the problem as I wasn't prepared to install "unnecessary" packages. But I have found that sometimes I have to do a dotnet build before a dotnet run or dotnet publish for Nuget to restore things correctly.

Yeah, installing those extra packages didn't feel right, but after spending 4 hours trying to figure it out, it's time to throw in the towel so I can move on with the project. 🤷‍♀️

I have been running "Rebuild Solution" in VS, then popping over to the browser with my custom hostname and refreshing.... So, I'd imagine that is equivalent to `dotnet build` ?
